Userland/Libraries/LibCore/IOSurface.cpp

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright (c) 2024, Aliaksandr Kalenik <kalenik.aliaksandr@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/OwnPtr.h>
+#include <LibCore/IOSurface.h>
+
+#if !defined(AK_OS_MACOS)
+static_assert(false, "This file must only be used for macOS");
+#endif
+
+#define FixedPoint FixedPointMacOS
+#define Duration DurationMacOS
+#include <IOSurface/IOSurface.h>
+#undef FixedPoint
+#undef Duration
+
+namespace Core {
+
+template<typename T>
+class RefAutoRelease {
+    AK_MAKE_NONCOPYABLE(RefAutoRelease);
+
+public:
+    RefAutoRelease(T ref)
+        : m_ref(ref)
+    {
+    }
+
+    ~RefAutoRelease()
+    {
+        if (m_ref)
+            CFRelease(m_ref);
+        m_ref = nullptr;
+    }
+
+    T ref() const { return m_ref; }
+
+private:
+    T m_ref { nullptr };
+};
+
+struct IOSurfaceHandle::IOSurfaceRefWrapper {
+    IOSurfaceRef ref;
+};
+
+IOSurfaceHandle::IOSurfaceHandle(OwnPtr<IOSurfaceRefWrapper>&& ref_wrapper)
+    : m_ref_wrapper(move(ref_wrapper))
+{
+}
+
+IOSurfaceHandle::IOSurfaceHandle(IOSurfaceHandle&& other) = default;
+IOSurfaceHandle& IOSurfaceHandle::operator=(IOSurfaceHandle&& other) = default;
+
+IOSurfaceHandle::~IOSurfaceHandle()
+{
+    if (m_ref_wrapper)
+        CFRelease(m_ref_wrapper->ref);
+}
+
+IOSurfaceHandle IOSurfaceHandle::create(int width, int height)
+{
+    size_t bytes_per_element = 4;
+    uint32_t pixel_format = 'BGRA';
+
+    RefAutoRelease<CFNumberRef> width_number = CFNumberCreate(kCFAllocatorDefault, kCFNumberIntType, &width);
+    RefAutoRelease<CFNumberRef> height_number = CFNumberCreate(kCFAllocatorDefault, kCFNumberIntType, &height);
+    RefAutoRelease<CFNumberRef> bytes_per_element_number = CFNumberCreate(kCFAllocatorDefault, kCFNumberIntType, &bytes_per_element);
+    RefAutoRelease<CFNumberRef> pixel_format_number = CFNumberCreate(kCFAllocatorDefault, kCFNumberSInt32Type, &pixel_format);
+
+    CFMutableDictionaryRef props = CFDictionaryCreateMutable(kCFAllocatorDefault,
+        0,
+        &kCFTypeDictionaryKeyCallBacks,
+        &kCFTypeDictionaryValueCallBacks);
+
+    CFDictionarySetValue(props, kIOSurfaceWidth, width_number.ref());
+    CFDictionarySetValue(props, kIOSurfaceHeight, height_number.ref());
+    CFDictionarySetValue(props, kIOSurfaceBytesPerElement, bytes_per_element_number.ref());
+    CFDictionarySetValue(props, kIOSurfacePixelFormat, pixel_format_number.ref());
+
+    auto* ref = IOSurfaceCreate(props);
+    VERIFY(ref);
+    return IOSurfaceHandle(make<IOSurfaceRefWrapper>(ref));
+}
+
+MachPort IOSurfaceHandle::create_mach_port(MachPort::PortRight port_right) const
+{
+    auto port = IOSurfaceCreateMachPort(m_ref_wrapper->ref);
+    return MachPort::adopt_right(port, port_right);
```

Comment:
No, this makes no sense :disappointed:

`IOSurfaceCreateMachPort()` always creates a send right, no matter what `MachPort::PortRight port_right` argument you would've passed to this method. So you should do this:

```suggestion
MachPort IOSurfaceHandle::create_mach_port() const
{
    auto port = IOSurfaceCreateMachPort(m_ref_wrapper->ref);
    return MachPort::adopt_right(port, MachPort::PortRight::Send);
```

`adopt_right` doesn't do any magic, it simply wraps a "raw" Mach port right (a C `mach_port_t`) into the C++ `MachPort` type. The right type passed in must match the actual type of right that the raw `mach_port_t` represents. So it must always be `Send`, since that's the type of right `IOSurfaceCreateMachPort()` returns, it makes no sense to ask the caller to supply the right type.

----

That @ADKaster would suggest this, and that you would write it like this, indicates that you guys are misunderstanding something fundamental about Mach ports/IPC (which only makes sense given how severely underdocumented Mach is!), but I don't know what exactly, 'cause you seem to have figured a lot of it out otherwise. So if anything's unclear / doesn't make sense / "click" about Mach IPC, please ask me, I'm here to help! :slightly_smiling_face:

Reply:
is there any place I could use as a reference that explicitly says `IOSurfaceCreateMachPort()` always returns send right?

Reply:
[Apple's docs](https://developer.apple.com/documentation/iosurface/iosurfacecreatemachport(_:)?language=objc) don't seem to say this anywhere, but:
1. Returning a send right is the natural thing to do.
2. You should be able to verify that what you're getting is a send right. And I don't mean in Ladybird at runtime; I mean, write a tiny program that creates an IOSurface, calls `IOSurfaceCreateMachPort()`, and checks (with `mach_port_type()`) which type it's getting.
3. (As I already mentioned yesterday), I assumed that it returns a send right when writing Owl, in particular I'm passing the returned port as a send right [here](https://github.com/owl-compositor/owl/blob/91abf02613cd2ddb97be58b5b6703240320233a0/Sources/Protocol/owl-iosurface-unstable-v1-mig.defs#L10). (`mach_port_t` in MIG corresponds to `MACH_MSG_TYPE_COPY_SEND`; a send-once right would be instead listed as `mach_port_move_send_once_t`.) Given that it worked, it must have been a send right indeed.

The type of right it returns certainly doesn't depend on the `MachPort::PortRight port_right` argument, if only because you're not passing that to `IOSurfaceCreateMachPort()`.